gensim/models/fasttext.py

```diff
@@ -0,0 +1,335 @@
+            if w != word.index:
+                word_indices.append(w)
+        l2b = model.syn1neg[word_indices]  # 2d matrix, k+1 x layer1_size
+        fb = expit(dot(l1, l2b.T))  # propagate hidden -> output
+        gb = (model.neg_labels - fb) * alpha  # vector of error gradients multiplied by the learning rate
+        if learn_hidden:
+            model.syn1neg[word_indices] += outer(gb, l1)  # learn hidden -> output
+        neu1e += dot(gb, l2b)  # save error
+
+    if learn_vectors:
+        # learn input -> hidden, here for all words in the window separately
+        if not model.cbow_mean and input_subword_indices:
+            neu1e /= len(input_subword_indices)
+        for i in input_subword_indices:
+            model.wv.syn0_all[i] += neu1e * model.syn0_all_lockf[i]
+
+    return neu1e
+
+def train_batch_sg(model, sentences, alpha, work=None):
+    result = 0
+    for sentence in sentences:
+        word_vocabs = [model.wv.vocab[w] for w in sentence if w in model.wv.vocab] #and
+                       # model.wv.vocab[w].sample_int > model.random.rand() * 2**32]
+        for pos, word in enumerate(word_vocabs):
+            # reduced_window = model.random.randint(model.window)  # `b` in the original word2vec code
+            # now go over all words from the (reduced) window, predicting each one in turn
+            # start = max(0, pos - model.window + reduced_window)
+            start = max(0, pos - model.window)
+            # for pos2, word2 in enumerate(word_vocabs[start:(pos + model.window + 1 - reduced_window)], start):
+
+            subwords_indices = [word.index]
+            word2_subwords = model.wv.ngrams_word[model.wv.index2word[word.index]]
+            # print("word2_subwords: ", word2_subwords)
+            for subword in word2_subwords:
+                subwords_indices.append(model.wv.ngrams[subword])
+
+            for pos2, word2 in enumerate(word_vocabs[start:(pos + model.window + 1)], start):
+                if pos2 != pos:  # don't train on the `word` itself
+                    # print("sending pair : ", model.wv.index2word[word2.index], " , ", model.wv.index2word[word.index])
+                    # subwords_indices = [word2.index]
+                    # word2_subwords = model.wv.ngrams_word[model.wv.index2word[word2.index]]
+
+
+                    # train_sg_pair(model, model.wv.index2word[word.index], subwords_indices, alpha)
+                    train_sg_pair(model, model.wv.index2word[word2.index], subwords_indices, alpha)
+
+        result += len(word_vocabs)
+    return result
+
+def train_sg_pair(model, word, input_subword_indices, alpha, learn_vectors=True, learn_hidden=True, context_vectors=None, context_locks=None):
+    if context_vectors is None:
+        context_vectors = model.wv.syn0_all
+    if context_locks is None:
+        context_locks = model.syn0_all_lockf
+
+    if word not in model.wv.vocab:
```

Comment:
Yes. That's correct. This was in the [Word2Vec code](https://github.com/RaRe-Technologies/gensim/blob/develop/gensim/models/word2vec.py#L263) as well so it slipped through I guess. Thanks for pointing this out!